components/renku_data_services/base_api/auth.py

```diff
@@ -45,3 +45,38 @@ async def decorated_function(request: Request, user: APIUser, *args, **kwargs):
         return response
 
     return decorated_function
+
+
+def only_authenticated(f):
+    """Decorator that errors out if the user is not authenticated.
+
+    It looks for APIUser in the named or unnamed poarameters.
+    """
+
+    @wraps(f)
+    async def decorated_function(self, *args, **kwargs):
+        import logging
+
+        logging.error(f"Receiving args: {args}, kwargs{kwargs}")
+        api_user = None
+        if "requested_by" in kwargs and isinstance(kwargs["requested_by"], APIUser):
+            api_user = kwargs["requested_by"]
+        elif "user" in kwargs and isinstance(kwargs["user"], APIUser):
+            api_user = kwargs["user"]
+        elif len(args) >= 1:
+            api_user_search = [a for a in args if isinstance(a, APIUser)]
+            if len(api_user_search) == 1:
+                api_user = api_user_search[0]
+            else:
+                raise errors.ProgrammingError(
+                    detail="Found two valid non-keyword APIUser arguments when authenticating user, expected only one."
+                )
+
+        if api_user is None or not api_user.is_authenticated:
+            raise errors.Unauthorized(message="You have to be authenticated to perform this operation.")
+
+        # the user is authenticated
+        response = await f(self, *args, **kwargs)
+        return response
```

Comment:
i wonder, can you just return the future instead of awaiting it and then the calling code has to await again?

Reply:
Without the await this breaks at the sanic level.

```
Traceback (most recent call last):
  File "/home/renku/.cache/pypoetry/virtualenvs/renku-data-services-tbxlFWds-py3.11/lib/python3.11/site-packages/sanic/request/types.py", line 347, in respond
    response = await self.app._run_response_middleware(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "_run_response_middleware", line 17, in _run_response_middleware
    from asyncio.futures import Future
                            ^^^^^^^^^^^
  File "/home/renku/.cache/pypoetry/virtualenvs/renku-data-services-tbxlFWds-py3.11/lib/python3.11/site-packages/sanic_ext/extensions/http/cors.py", line 56, in _add_cors_headers
    if ORIGIN_HEADER not in response.headers:
                            ^^^^^^^^^^^^^^^^
AttributeError: 'coroutine' object has no attribute 'headers'
```

But I am sure we can work something out - maybe with middleware that checks for this case and awaits? Or maybe there is a nicer way. But simple removing the await does not work.

Reply:
I would just do this in a follow up issue.

Reply:
oh it doesn't matter, i just thought that futures can be passed around without awaiting